mdata/reorder_buffer.go

```diff
@@ -5,50 +5,52 @@ import (
 )
 
 // ReorderBuffer keeps a window of data during which it is ok to send data out of order.
-// Once the reorder window has passed Add() will return the old data and delete it from the buffer.
 // The reorder buffer itself is not thread safe because it is only used by AggMetric,
 // which is thread safe, so there is no locking in the buffer.
+//
+// newest=0 may mean no points added yet, or newest point is at position 0.
+// we use the Ts of points in the buffer to check for valid points. Ts == 0 means no point
+// in particular newest.Ts == 0 means the buffer is empty
+// the buffer is evenly spaced (points are `interval` apart) and may be sparsely populated
 type ReorderBuffer struct {
-	len      uint32         // length of buffer in number of datapoints
 	newest   uint32         // index of newest buffer entry
 	interval uint32         // metric interval
 	buf      []schema.Point // the actual buffer holding the data
 }
 
 func NewReorderBuffer(reorderWindow uint32, interval int) *ReorderBuffer {
-	buf := &ReorderBuffer{
-		len:      reorderWindow,
+	return &ReorderBuffer{
 		interval: uint32(interval),
 		buf:      make([]schema.Point, reorderWindow),
 	}
-	return buf
 }
 
+// Add adds the point if it falls within the window.
+// it returns points that have been purged out of the buffer, as well as whether the add succeeded.
 func (rob *ReorderBuffer) Add(ts uint32, val float64) ([]schema.Point, bool) {
 	ts = AggBoundary(ts, rob.interval)
 
 	// out of order and too old
-	if rob.buf[rob.newest].Ts != 0 && ts <= rob.buf[rob.newest].Ts-(rob.len*rob.interval) {
+	if rob.buf[rob.newest].Ts != 0 && ts <= rob.buf[rob.newest].Ts-(uint32(cap(rob.buf))*rob.interval) {
```

Comment:
I think it's a good idea to use `cap()` instead of having to store `.len` as an additional attribute. I'm not sure how expensive/cheap the `uint32()` conversion is, but wouldn't it be nicer to do something like `len := uint32(cap(rob.buf))` in the beginning of `Add()` then reuse that var?

Reply:
> I'm not sure how expensive/cheap the uint32() conversion is

see the benchmark results in the commit messages. there's no significant change